### PR TITLE
Add rlang type checkers

### PR DESCRIPTION
**Pull request by topepo**
Does some additional tests for basic things. The next step would be adding these checks to the steps.

### CODE REVIEW

File: R/steps_and_checks.R [hunk 1]
```diff
@@ -5,24 +5,41 @@
...
 #' @rdname step
 #' @export
-check <- function(subclass, ..., .prefix = "check_") {
-  structure(list(...),
+check <- function(subclass, ..., .prefix = "check_",
+                  call = rlang::caller_env()) {
+  args <- list(...)
+
+  check_string(subclass, call = call)
+  .prefix <- rlang::arg_match0(.prefix, c("step_", "check_", ""),
```
Comment (topepo) [inline R/steps_and_checks.R:38]: We do have some code in `update()` that assigns a prefix of `""` and I didn't want to mess with that.

File: R/steps_and_checks.R [hunk 2]
```diff
@@ -51,3 +68,22 @@ add_check <- function(rec, object) {
   rec$steps[[length(rec$steps) + 1]] <- object
   rec
 }
+
+# ------------------------------------------------------------------------------
+
+check_step_check_args <- function(x, call = rlang::caller_env()) {
```
Comment (EmilHvitfeldt) [inline R/steps_and_checks.R:74]: i like this idea! makes it very clean. what about checking for argument used a lot but not everywhere. like `keep_original_cols`?

File: R/steps_and_checks.R [hunk 3]
```diff
@@ -5,24 +5,41 @@
...
     class = c(paste0(.prefix, subclass), "step")
   )
 }
 
 #' @rdname step
 #' @export
-check <- function(subclass, ..., .prefix = "check_") {
-  structure(list(...),
+check <- function(subclass, ..., .prefix = "check_",
+                  call = rlang::caller_env()) {
```
Comment (EmilHvitfeldt) [inline R/steps_and_checks.R:34]:
```suggestion
check <- function(subclass, ..., .prefix = "check_",
                  call = rlang::caller_env(2)) {
```

gives a better message

Screenshot 2024-10-24 at 12 46 43 PM

File: R/steps_and_checks.R [hunk 4]
```diff
@@ -5,24 +5,41 @@
...
 #'
 #' @seealso [developer_functions]
 #'
 #' @keywords internal
 #' @return An updated step or check with the new class.
 #' @export
-step <- function(subclass, ..., .prefix = "step_") {
-  structure(list(...),
+step <- function(subclass, ..., .prefix = "step_",
+                 call = rlang::caller_env()) {
```
Comment (EmilHvitfeldt) [inline R/steps_and_checks.R:18]:
```suggestion
step <- function(subclass, ..., .prefix = "step_",
                 call = rlang::caller_env(2)) {
```

File: tests/testthat/test-basics.R [hunk 5]
```diff
@@ -426,3 +426,37 @@ test_that("data argument is checked in recipe.formula() (#1325)", {
...
+      step(
+        subclass = "lightly",
+        trained = trained,
+        skip = skip,
+        id = id
+      )
+    }
+
+  expect_snapshot(
+    step_lightly(trained = "yes"),
```
Comment (EmilHvitfeldt) [inline tests/testthat/test-basics.R:443]: top-tier function name. but it doesn't reflect the call structure of a step, as this is what `step_lightly_new()` looks like.

we can instead use something like

```suggestion
    recipe(~., mtcars) %>% step_normalize(trained = "yes"),
```

**Review by EmilHvitfeldt — APPROVED** [review]
overall like what is happening. good idea to check for common arguments in `step()` and `check()` as all the steps run through there. 

I want to see what happens if we extend it to a few common arguments like `keep_original_cols` and `case_weights` that has very specific meanings, but aren't used all the time

**Comment by EmilHvitfeldt** [comment]
Here are the most common arguments for recipes steps. So I'll go back on my word before, and way we shouldn't check for `case_weights`

``` r
library(recipes)
library(tidyverse)

fns <- lsf.str("package:recipes")
fn_args <- lapply(fns, formals)
names(fn_args) <- fns

fn_args <- fn_args[str_detect(fns, "^step_") & str_detect(fns, "_new", negate = TRUE)]

map(fn_args, ~ tibble(arg = names(.x))) |>
  list_rbind(names_to = "step") |>
  count(arg, sort = TRUE) |>
  print(n = 20)
#> # A tibble: 127 × 2
#>    arg                    n
#>    <chr>              <int>
#>  1 id                    98
#>  2 recipe                98
#>  3 role                  98
#>  4 skip                  98
#>  5 trained               98
#>  6 ...                   96
#>  7 columns               36
#>  8 keep_original_cols    36
#>  9 options               26
#> 10 prefix                16
#> 11 objects               11
#> 12 res                    9
#> 13 threshold              9
#> 14 degree                 8
#> 15 num_comp               8
#> 16 deg_free               7
#> 17 inputs                 7
#> 18 complete_set           6
#> 19 levels                 6
#> 20 na_rm                  6
#> # ℹ 107 more rows
```

<sup>Created on 2024-10-24 with [reprex v2.1.0](https://reprex.tidyverse.org)</sup>

**Comment by topepo** [comment]
Extending that to extension packages: 

``` r
suppressPackageStartupMessages(library(tidymodels))
suppressPackageStartupMessages(library(revdepcheck))

recipe_pkgs <- revdepcheck::cran_revdeps("recipes", dependencies = c("Depends", "Imports"))
recipe_pkgs <- c(recipe_pkgs, "recipes")

recipe_pkgs <- sort(unique(c(recipe_pkgs)))
excl <- c("hydrorecipes", "healthcareai", "viruslearner")
recipe_pkgs <- recipe_pkgs[!(recipe_pkgs %in% excl)]

# pak::pak(recipe_pkgs, ask = FALSE)

excl_internal <- c("check_type", "check_nominal_type", "check_factor_vars", 
                   "check_name", "check_is_lat_lon", "check_new_data", 
                   "check_role_requirements", "check_bake_role_requirements", 
                   "check_step_check_args")         
        
listify <- function(x, nm) {
  tibble::tibble(fn = nm, argument = names(x))
} 

get_steps_args <- function(pkg, excl = character(0)) {
  suppressPackageStartupMessages(library(pkg, character.only = TRUE))
  fns <- lsf.str(paste0("package:", pkg), pattern = ("(step_)|(check_)"))
  fns <- fns[!(fns %in% excl_internal)]
  fn_args <- lapply(fns, formals)
  purrr::map2_dfr(fn_args, fns, listify) %>% dplyr::mutate(package = pkg) %>% relocate(package)
} 

pkg_arg_tbl <- 
  map_dfr(recipe_pkgs, get_steps_args) %>% 
  anti_join(tibble::tibble(package = "recipes", fn = excl_internal))
#> Joining with `by = join_by(package, fn)`

pkg_arg_tbl %>% count(argument) %>% arrange(desc(n))
#> # A tibble: 295 × 2
#>    argument               n
#>    <chr>              <int>
#>  1 ...                  216
#>  2 id                   199
#>  3 recipe               199
#>  4 role                 199
#>  5 skip                 199
#>  6 trained              194
#>  7 columns               90
#>  8 keep_original_cols    57
#>  9 options               41
#> 10 prefix                39
#> # ℹ 285 more rows
```

<sup>Created on 2024-10-25 with [reprex v2.1.0](https://reprex.tidyverse.org)</sup>

Maybe add `role`?

**Comment by EmilHvitfeldt** [comment]
yes, that sounds reasonable!

**Review by EmilHvitfeldt — APPROVED** [review]
amazing!

**Comment by topepo** [comment]
Now on to part two!

**Pull request merged**